File: mobile/src/main/java/com/backdoor/moove/core/fragments/OtherSettingsFragment.java
package com.backdoor.moove.core.fragments;

import android.app.AlertDialog;
import android.content.ActivityNotFoundException;
import android.content.Context;
import android.content.Intent;
import android.content.pm.PackageInfo;
import android.content.pm.PackageManager;
import android.net.Uri;
import android.os.Bundle;
import androidx.annotation.NonNull;
import androidx.fragment.app.Fragment;
import androidx.appcompat.app.ActionBar;
import androidx.appcompat.app.AppCompatActivity;
import android.view.LayoutInflater;
import android.view.View;
import android.view.ViewGroup;
import android.widget.ArrayAdapter;
import android.widget.LinearLayout;
import android.widget.TextView;
import android.widget.Toast;

import com.backdoor.moove.R;
import com.backdoor.moove.core.consts.Configs;
import com.backdoor.moove.core.dialogs.ChangeDialog;
import com.backdoor.moove.core.dialogs.Help;
import com.backdoor.moove.core.dialogs.ThanksDialog;
import com.backdoor.moove.core.helper.Module;
import com.backdoor.moove.core.helper.Permissions;
import com.backdoor.moove.databinding.DialogAboutLayoutBinding;

import java.util.ArrayList;
import java.util.List;

public class OtherSettingsFragment extends Fragment {

    private ActionBar ab;
    private List<Item> mDataList = new ArrayList<>();

    @Override
    public View onCreateView(LayoutInflater inflater, ViewGroup container, Bundle savedInstanceState) {
        View rootView = inflater.inflate(R.layout.settings_other, container, false);
        ab = ((AppCompatActivity) getActivity()).getSupportActionBar();
        if (ab != null) {
            ab.setTitle(R.string.other);
        }
        if (Module.isLollipop()) {
            rootView.findViewById(R.id.otherCard).setElevation(Configs.CARD_ELEVATION);
        }
        TextView about = rootView.findViewById(R.id.about);
        about.setOnClickListener(v -> showAboutDialog());

        TextView changes = rootView.findViewById(R.id.changes);
        changes.setOnClickListener(v -> getActivity().getApplicationContext()
                .startActivity(new Intent(getActivity().getApplicationContext(),
                        ChangeDialog.class)
                        .addFlags(Intent.FLAG_ACTIVITY_NEW_TASK)));

        TextView rateApp = rootView.findViewById(R.id.rateApp);
        rateApp.setOnClickListener(v -> launchMarket(getActivity()));

        TextView thanks = rootView.findViewById(R.id.thanks);
        thanks.setOnClickListener(v -> getActivity().getApplicationContext()
                .startActivity(new Intent(getActivity().getApplicationContext(),
                        ThanksDialog.class)
                        .addFlags(Intent.FLAG_ACTIVITY_NEW_TASK)));

        TextView help = rootView.findViewById(R.id.help);
        help.setOnClickListener(v -> getActivity().getApplicationContext()
                .startActivity(new Intent(getActivity().getApplicationContext(),
                        Help.class)
                        .addFlags(Intent.FLAG_ACTIVITY_NEW_TASK)));

        TextView menuFeedback = rootView.findViewById(R.id.menuFeedback);
        menuFeedback.setOnClickListener(v -> {
            final Intent emailIntent = new Intent(Intent.ACTION_SEND);
            emailIntent.setType("plain/text");
            emailIntent.putExtra(Intent.EXTRA_EMAIL,
                    new String[]{"feedback.cray@gmail.com"});
            emailIntent.putExtra(Intent.EXTRA_SUBJECT, "Moove");
            getActivity().startActivity(Intent.createChooser(emailIntent, "Send mail..."));
        });

        TextView menuShare = rootView.findViewById(R.id.menuShare);
        menuShare.setOnClickListener(v -> {
            Intent shareIntent = new Intent(Intent.ACTION_SEND);
            shareIntent.setType("text/plain");
            shareIntent.putExtra(Intent.EXTRA_TEXT,
                    "https://play.google.com/store/apps/details?id=" +
                            getActivity().getPackageName());
            getActivity().startActivity(Intent.createChooser(shareIntent, "Share..."));
        });

        LinearLayout permissionBlock = rootView.findViewById(R.id.permissionBlock);
        if (Module.isMarshmallow()) {
            permissionBlock.setVisibility(View.VISIBLE);
        } else {
            permissionBlock.setVisibility(View.GONE);
        }

        TextView permissionExplain = rootView.findViewById(R.id.permissionExplain);
        permissionExplain.setOnClickListener(v -> getActivity().getApplicationContext()
                .startActivity(new Intent(getActivity().getApplicationContext(), ThanksDialog.class)
                        .putExtra("int", 1)
                        .addFlags(Intent.FLAG_ACTIVITY_NEW_TASK)));

        TextView permissionEnable = rootView.findViewById(R.id.permissionEnable);
        permissionEnable.setOnClickListener(v -> showPermissionDialog());

        return rootView;
    }

    private void launchMarket(Context context) {
        Uri uri = Uri.parse("market://details?id=" + context.getPackageName());
        Intent goToMarket = new Intent(Intent.ACTION_VIEW, uri);
        try {
            context.startActivity(goToMarket);
        } catch (ActivityNotFoundException e) {
            Toast.makeText(context, context.getString(R.string.could_not_launch_market), Toast.LENGTH_SHORT).show();
        }
    }

    @Override
    public void onDetach() {
        super.onDetach();
        ab = ((AppCompatActivity) getActivity()).getSupportActionBar();
        if (ab != null) {
            ab.setTitle(R.string.settings);
        }
    }

    private void showAboutDialog() {
        AlertDialog.Builder builder = new AlertDialog.Builder(getContext());
        DialogAboutLayoutBinding binding = DialogAboutLayoutBinding.inflate(LayoutInflater.from(getContext()));
        PackageInfo pInfo;
        try {
            pInfo = getContext().getPackageManager().getPackageInfo(getContext().getPackageName(), 0);
            binding.appVersion.setText(pInfo.versionName);
        } catch (PackageManager.NameNotFoundException e) {
            e.printStackTrace();
        }
        builder.setView(binding.getRoot());
        builder.create().show();
    }

    private void showPermissionDialog() {
        if (!loadDataToList()) return;
        AlertDialog.Builder builder = new AlertDialog.Builder(getContext());
        builder.setTitle(R.string.allow_permission);
        builder.setSingleChoiceItems(new ArrayAdapter<Item>(getContext(), android.R.layout.simple_list_item_1) {
            @Override
            public int getCount() {
                return mDataList.size();
            }

            @NonNull
            @Override
            public View getView(int position, View convertView, @NonNull ViewGroup parent) {
                if (convertView == null) {
                    convertView = LayoutInflater.from(getContext()).inflate(android.R.layout.simple_list_item_1, parent, false);
                }
                TextView tvName = convertView.findViewById(android.R.id.text1);
                tvName.setText(mDataList.get(position).getTitle());
                return convertView;
            }
        }, -1, (dialogInterface, i) -> {
            dialogInterface.dismiss();
            requestPermission(i);
        });
        builder.create().show();
    }

    private boolean loadDataToList() {
        mDataList.clear();
        if (!Permissions.checkPermission(getActivity(), Permissions.ACCESS_COARSE_LOCATION)) {
            mDataList.add(new Item(getString(R.string.course_location), Permissions.ACCESS_COARSE_LOCATION));
        }
        if (!Permissions.checkPermission(getActivity(), Permissions.ACCESS_FINE_LOCATION)) {
            mDataList.add(new Item(getString(R.string.fine_location), Permissions.ACCESS_FINE_LOCATION));
        }
        if (!Permissions.checkPermission(getActivity(), Permissions.CALL_PHONE)) {
            mDataList.add(new Item(getString(R.string.call_phone), Permissions.CALL_PHONE));
        }
        if (!Permissions.checkPermission(getActivity(), Permissions.GET_ACCOUNTS)) {
            mDataList.add(new Item(getString(R.string.get_accounts), Permissions.GET_ACCOUNTS));
        }
        if (!Permissions.checkPermission(getActivity(), Permissions.READ_PHONE_STATE)) {
            mDataList.add(new Item(getString(R.string.read_phone_state), Permissions.READ_PHONE_STATE));
        }
        if (!Permissions.checkPermission(getActivity(), Permissions.READ_CONTACTS)) {
            mDataList.add(new Item(getString(R.string.read_contacts), Permissions.READ_CONTACTS));
        }
        if (!Permissions.checkPermission(getActivity(), Permissions.READ_EXTERNAL)) {
            mDataList.add(new Item(getString(R.string.read_external_storage), Permissions.READ_EXTERNAL));
        }
        if (!Permissions.checkPermission(getActivity(), Permissions.WRITE_EXTERNAL)) {
            mDataList.add(new Item(getString(R.string.write_external_storage), Permissions.WRITE_EXTERNAL));
        }
        if (!Permissions.checkPermission(getActivity(), Permissions.SEND_SMS)) {
            mDataList.add(new Item(getString(R.string.send_sms), Permissions.SEND_SMS));
        }
        if (mDataList.size() == 0) {
            Toast.makeText(getContext(), R.string.all_permissions_are_enabled, Toast.LENGTH_SHORT).show();
            return false;
        } else {
            return true;
        }
    }

    private void requestPermission(int position) {
        requestPermissions(new String[]{mDataList.get(position).getPermission()}, 155);
    }

    @Override
    public void onRequestPermissionsResult(int requestCode, @NonNull String[] permissions, @NonNull int[] grantResults) {
        if (grantResults.length == 0) return;
        if (grantResults[0] == PackageManager.PERMISSION_GRANTED) {
            showPermissionDialog();
        }
    }

    private class Item {
        private String title, permission;

        Item(String title, String permission) {
            this.permission = permission;
            this.title = title;
        }

        public String getTitle() {
            return title;
        }

        public String getPermission() {
            return permission;
        }
    }
}

File: mobile/src/main/java/com/backdoor/moove/core/fragments/OtherSettingsFragment.kt
package com.backdoor.moove.core.fragments

import android.app.AlertDialog
import android.content.ActivityNotFoundException
import android.content.Context
import android.content.Intent
import android.content.pm.PackageInfo
import android.content.pm.PackageManager
import android.net.Uri
import android.os.Bundle
import androidx.fragment.app.Fragment
import androidx.appcompat.app.ActionBar
import androidx.appcompat.app.AppCompatActivity
import android.view.LayoutInflater
import android.view.View
import android.view.ViewGroup
import android.widget.ArrayAdapter
import android.widget.LinearLayout
import android.widget.TextView
import android.widget.Toast

import com.backdoor.moove.R
import com.backdoor.moove.core.consts.Configs
import com.backdoor.moove.core.dialogs.ChangeDialog
import com.backdoor.moove.core.dialogs.Help
import com.backdoor.moove.core.dialogs.ThanksDialog
import com.backdoor.moove.core.helper.Module
import com.backdoor.moove.core.helper.Permissions
import com.backdoor.moove.databinding.DialogAboutLayoutBinding

import java.util.ArrayList

class OtherSettingsFragment : Fragment() {

    private var ab: ActionBar? = null
    private val mDataList = ArrayList<Item>()

    override fun onCreateView(inflater: LayoutInflater, container: ViewGroup?, savedInstanceState: Bundle?): View? {
        val rootView = inflater.inflate(R.layout.settings_other, container, false)
        ab = (activity as AppCompatActivity).supportActionBar
        if (ab != null) {
            ab!!.setTitle(R.string.other)
        }
        if (Module.isLollipop) {
            rootView.findViewById<View>(R.id.otherCard).elevation = Configs.CARD_ELEVATION
        }
        val about = rootView.findViewById<TextView>(R.id.about)
        about.setOnClickListener { v -> showAboutDialog() }

        val changes = rootView.findViewById<TextView>(R.id.changes)
        changes.setOnClickListener { v ->
            activity!!.applicationContext
                    .startActivity(Intent(activity!!.applicationContext,
                            ChangeDialog::class.java)
                            .addFlags(Intent.FLAG_ACTIVITY_NEW_TASK))
        }

        val rateApp = rootView.findViewById<TextView>(R.id.rateApp)
        rateApp.setOnClickListener { v -> launchMarket(activity!!) }

        val thanks = rootView.findViewById<TextView>(R.id.thanks)
        thanks.setOnClickListener { v ->
            activity!!.applicationContext
                    .startActivity(Intent(activity!!.applicationContext,
                            ThanksDialog::class.java)
                            .addFlags(Intent.FLAG_ACTIVITY_NEW_TASK))
        }

        val help = rootView.findViewById<TextView>(R.id.help)
        help.setOnClickListener { v ->
            activity!!.applicationContext
                    .startActivity(Intent(activity!!.applicationContext,
                            Help::class.java)
                            .addFlags(Intent.FLAG_ACTIVITY_NEW_TASK))
        }

        val menuFeedback = rootView.findViewById<TextView>(R.id.menuFeedback)
        menuFeedback.setOnClickListener { v ->
            val emailIntent = Intent(Intent.ACTION_SEND)
            emailIntent.type = "plain/text"
            emailIntent.putExtra(Intent.EXTRA_EMAIL,
                    arrayOf("feedback.cray@gmail.com"))
            emailIntent.putExtra(Intent.EXTRA_SUBJECT, "Moove")
            activity!!.startActivity(Intent.createChooser(emailIntent, "Send mail..."))
        }

        val menuShare = rootView.findViewById<TextView>(R.id.menuShare)
        menuShare.setOnClickListener { v ->
            val shareIntent = Intent(Intent.ACTION_SEND)
            shareIntent.type = "text/plain"
            shareIntent.putExtra(Intent.EXTRA_TEXT,
                    "https://play.google.com/store/apps/details?id=" + activity!!.packageName)
            activity!!.startActivity(Intent.createChooser(shareIntent, "Share..."))
        }

        val permissionBlock = rootView.findViewById<LinearLayout>(R.id.permissionBlock)
        if (Module.isMarshmallow) {
            permissionBlock.visibility = View.VISIBLE
        } else {
            permissionBlock.visibility = View.GONE
        }

        val permissionExplain = rootView.findViewById<TextView>(R.id.permissionExplain)
        permissionExplain.setOnClickListener { v ->
            activity!!.applicationContext
                    .startActivity(Intent(activity!!.applicationContext, ThanksDialog::class.java)
                            .putExtra("int", 1)
                            .addFlags(Intent.FLAG_ACTIVITY_NEW_TASK))
        }

        val permissionEnable = rootView.findViewById<TextView>(R.id.permissionEnable)
        permissionEnable.setOnClickListener { v -> showPermissionDialog() }

        return rootView
    }

    private fun launchMarket(context: Context) {
        val uri = Uri.parse("market://details?id=" + context.packageName)
        val goToMarket = Intent(Intent.ACTION_VIEW, uri)
        try {
            context.startActivity(goToMarket)
        } catch (e: ActivityNotFoundException) {
            Toast.makeText(context, context.getString(R.string.could_not_launch_market), Toast.LENGTH_SHORT).show()
        }

    }

    override fun onDetach() {
        super.onDetach()
        ab = (activity as AppCompatActivity).supportActionBar
        if (ab != null) {
            ab!!.setTitle(R.string.settings)
        }
    }

    private fun showAboutDialog() {
        val builder = AlertDialog.Builder(context)
        val binding = DialogAboutLayoutBinding.inflate(LayoutInflater.from(context))
        val pInfo: PackageInfo
        try {
            pInfo = context!!.packageManager.getPackageInfo(context!!.packageName, 0)
            binding.appVersion.text = pInfo.versionName
        } catch (e: PackageManager.NameNotFoundException) {
            e.printStackTrace()
        }

        builder.setView(binding.root)
        builder.create().show()
    }

    private fun showPermissionDialog() {
        if (!loadDataToList()) return
        val builder = AlertDialog.Builder(context)
        builder.setTitle(R.string.allow_permission)
        builder.setSingleChoiceItems(object : ArrayAdapter<Item>(context!!, android.R.layout.simple_list_item_1) {
            override fun getCount(): Int {
                return mDataList.size
            }

            override fun getView(position: Int, convertView: View?, parent: ViewGroup): View {
                var convertView = convertView
                if (convertView == null) {
                    convertView = LayoutInflater.from(context).inflate(android.R.layout.simple_list_item_1, parent, false)
                }
                val tvName = convertView!!.findViewById<TextView>(android.R.id.text1)
                tvName.text = mDataList[position].title
                return convertView
            }
        }, -1) { dialogInterface, i ->
            dialogInterface.dismiss()
            requestPermission(i)
        }
        builder.create().show()
    }

    private fun loadDataToList(): Boolean {
        mDataList.clear()
        if (!Permissions.checkPermission(activity, Permissions.ACCESS_COARSE_LOCATION)) {
            mDataList.add(Item(getString(R.string.course_location), Permissions.ACCESS_COARSE_LOCATION))
        }
        if (!Permissions.checkPermission(activity, Permissions.ACCESS_FINE_LOCATION)) {
            mDataList.add(Item(getString(R.string.fine_location), Permissions.ACCESS_FINE_LOCATION))
        }
        if (!Permissions.checkPermission(activity, Permissions.CALL_PHONE)) {
            mDataList.add(Item(getString(R.string.call_phone), Permissions.CALL_PHONE))
        }
        if (!Permissions.checkPermission(activity, Permissions.GET_ACCOUNTS)) {
            mDataList.add(Item(getString(R.string.get_accounts), Permissions.GET_ACCOUNTS))
        }
        if (!Permissions.checkPermission(activity, Permissions.READ_PHONE_STATE)) {
            mDataList.add(Item(getString(R.string.read_phone_state), Permissions.READ_PHONE_STATE))
        }
        if (!Permissions.checkPermission(activity, Permissions.READ_CONTACTS)) {
            mDataList.add(Item(getString(R.string.read_contacts), Permissions.READ_CONTACTS))
        }
        if (!Permissions.checkPermission(activity, Permissions.READ_EXTERNAL)) {
            mDataList.add(Item(getString(R.string.read_external_storage), Permissions.READ_EXTERNAL))
        }
        if (!Permissions.checkPermission(activity, Permissions.WRITE_EXTERNAL)) {
            mDataList.add(Item(getString(R.string.write_external_storage), Permissions.WRITE_EXTERNAL))
        }
        if (!Permissions.checkPermission(activity, Permissions.SEND_SMS)) {
            mDataList.add(Item(getString(R.string.send_sms), Permissions.SEND_SMS))
        }
        if (mDataList.size == 0) {
            Toast.makeText(context, R.string.all_permissions_are_enabled, Toast.LENGTH_SHORT).show()
            return false
        } else {
            return true
        }
    }

    private fun requestPermission(position: Int) {
        requestPermissions(arrayOf(mDataList[position].permission), 155)
    }

    override fun onRequestPermissionsResult(requestCode: Int, permissions: Array<String>, grantResults: IntArray) {
        if (grantResults.size == 0) return
        if (grantResults[0] == PackageManager.PERMISSION_GRANTED) {
            showPermissionDialog()
        }
    }

    private inner class Item internal constructor(val title: String, val permission: String)
}
